Add fading sounds

SmallJoker opened the pull request:
This is a revised pull of #4304.
It adds (Lua) API functions to fade sounds.

EDIT: Tested in singleplayer, works as expected.

SmallJoker commented on src/network/clientpackethandler.cpp:
```diff
@@ -766,21 +766,40 @@ void Client::handleCommand_ItemDef(NetworkPacket* pkt)
 	*pkt >> server_id >> name >> gain >> type >> pos >> object_id >> loop;
 
+	// Check for remaining data in this packet
+	u32 packet_remaining = pkt->getSize() - 26 - name.length();
```
@nerzhul Is this possible or is the packet size undefined here?

sfan5 replied:
Just use
```c++
	try {
		*pkt >> fade;
	} catch(SerializationError &e) {};
```

nerzhul replied:
yeah if it's a new parameter, just use the try catch method

sfan5 commented on src/network/clientpackethandler.cpp:
```diff
@@ -766,21 +766,40 @@ void Client::handleCommand_ItemDef(NetworkPacket* pkt)
 {
+	/*
+		[0] u32 server_id
+		[4] u16 + ? string name
```
<pre>
[4] u16 name length
[6] char name[len]
</pre>

nerzhul commented on src/network/clientpackethandler.cpp:
```diff
@@ -819,6 +838,23 @@ void Client::handleCommand_StopSound(NetworkPacket* pkt)
 
+void Client::handleCommand_FadeSound(NetworkPacket *pkt)
+{
+	s32 server_id;
```
sound id is better than server_id, it's not a server

nerzhul commented on src/server.cpp:
```diff
@@ -2092,6 +2093,27 @@ void Server::stopSound(s32 handle)
+	UNORDERED_MAP<s32, ServerPlayingSound>::iterator i =
+			m_playing_sounds.find(handle);
+	if (i == m_playing_sounds.end())
+		return;
```
can you add an empty line after ?

SmallJoker replied:
Added.

nerzhul commented on src/sound_openal.cpp:
```diff
@@ -572,6 +596,45 @@ class OpenALSoundManager: public ISoundManager
+	void fadeSound(int soundid, float step, float gain)
+	{
+		float cGain = getSoundGain(soundid);
+		FadeState f = FadeState(step, cGain, gain);
```
you don't need for a temporary variable here

nerzhul commented on src/network/clientpackethandler.cpp:
```diff
@@ -766,21 +766,39 @@ void Client::handleCommand_ItemDef(NetworkPacket* pkt)
 
+	try {
+		*pkt >> fade;
+	} catch(SerializationError &e) {};
```
missing space before catch

nerzhul commented on src/sound_openal.cpp:
```diff
@@ -572,6 +596,44 @@ class OpenALSoundManager: public ISoundManager
+
+	void fadeSound(int soundid, float step, float gain)
+	{
+		FadeState f = FadeState(step, getSoundGain(soundid), gain);
```
no need for a temp variable here

nerzhul commented on src/network/clientpackethandler.cpp:
```diff
@@ -766,21 +766,39 @@ void Client::handleCommand_ItemDef(NetworkPacket* pkt)
 	v3f pos;
 	u16 object_id;
 	bool loop;
+	float fade = 0;
```
0.0f please

nerzhul commented on doc/lua_api.txt:
```diff
@@ -456,11 +456,13 @@ Examples of sound parameter tables:
     -- Play locationless on all clients
     {
         gain = 1.0, -- default
+        fade = 0, -- default, change to a value > 0 to fade the sound in
```
it's a float, please use 0.0 in your doc

nerzhul commented on src/network/clientpackethandler.cpp:
```diff
@@ -766,21 +766,39 @@ void Client::handleCommand_ItemDef(NetworkPacket* pkt)
 	v3f pos;
 	u16 object_id;
 	bool loop;
+	float fade = 0.f;
```
0.0f please to respect the standard codestyle we have on float

nerzhul commented on src/server.h:
```diff
@@ -123,7 +124,8 @@ struct ServerSoundParams
 		max_hear_distance(32*BS),
-		loop(false)
+		loop(false),
+		fade(0)
```
same thing on constructor

SmallJoker replied:
So is `gain` , `pos` and `max_hear_distance` initialized the wrong way too?
Could we leave this untouched and only change it for calculations/comparisons?

nerzhul commented on src/sound_openal.cpp:
```diff
@@ -572,6 +596,43 @@ class OpenALSoundManager: public ISoundManager
+		static float fadeDelay = 0;
+		fadeDelay += dtime;
+
+		if (fadeDelay < 0.2)
```
0.2f

nerzhul commented on src/sound_openal.cpp:
```diff
@@ -572,6 +596,43 @@ class OpenALSoundManager: public ISoundManager
+
+	void doFades(float dtime)
+	{
+		static float fadeDelay = 0;
```
why is it static ?

SmallJoker replied:
I suppose this was made static because it has no other use in this class.

nerzhul replied:
maybe it should be a static private class member instead

nerzhul commented on src/sound_openal.cpp:
```diff
@@ -572,6 +596,43 @@ class OpenALSoundManager: public ISoundManager
+				m_sounds_fading.erase(i->first);
+			}
+		}
+		fadeDelay = 0;
```
please remove static from fadeDelay this line will become useless

nerzhul commented on src/network/networkprotocol.h:
```diff
@@ -605,6 +605,13 @@ enum ToClientCommand
+		float step
+		float gain
+	 */
+
```
a protocol bump is required

sofar commented:
I'd love to see this in 0.4.16. It's been around too long already.

sofar commented:
Aargh I rebased this (see sofar/minetest:fade) but

- fade out works ok choppy, though :( 
- fade in is completely b0rk, crashes the game if `fade` is 0.01 or higher

@Ekdohibs and me are looking at crashes in the iteratot since `erase` is called inside the iterator loop:

Crash happens here:

https://github.com/minetest/minetest/pull/5340/files#diff-47d309d83e84f8ef92ac99f872441669R620

due to this line?

https://github.com/minetest/minetest/pull/5340/files#diff-47d309d83e84f8ef92ac99f872441669R634

SmallJoker commented:
@sofar Thank you so much for pointing this out! I rebased this pull and changed the `erase` calls in `Server::fadeSound` and `doFades` so that they do not mess up the loop anymore.
I'll do some backwards compability tests in the next days.

Ekdohibs commented on src/sound_openal.cpp:
```diff
@@ -567,16 +595,55 @@ class OpenALSoundManager: public ISoundManager
+				if (i->second.target_gain <= 0.f)
+					stopSound(i->first);
+
+				m_sounds_fading.erase(i++);
```
Just pointing out: with `std::unordered_map` (so in C++11), this is completely correct only with C++14. With C++11, it will not crash, however there is no guarantee that all sounds will be updated once and only once. As this is not very bad, and classic compilers will have the standard library required by C++14 anyway, I don't think that's a problem -- however, this is something that need to be kept in mind when this iteration is more important (I'm wondering if we're not using such a pattern elsewhere in the codebase anyway).

SmallJoker replied:
We're already using the same pattern in multiple places:
https://github.com/minetest/minetest/blob/master/src/serverenvironment.cpp#L1444
https://github.com/minetest/minetest/blob/master/src/nodedef.cpp#L1330
`std::map` only https://github.com/minetest/minetest/blob/master/src/particles.cpp#L448
And possibly a few more - haven't gone through all search results. Thanks for the review :)

sofar commented:
Fully tested by making a sound and continuously fading it in and out with 1 handle -> OK

One important notice is that if you start a local server *and* immediately start playing a sound to a local player or a single player, it *never* fades in. You have to wait until the client is fully started before the sound can be faded in. I was unable to play a sound inside a `register_onjoinplayer` that had a 0.0 gain or fade it in right there. But, I consider that a minor issue and not critical.

fade values of 0.2 correspond logically with 5 second fades from 0.0 to 1.0 and vice versa (-0.2), which is logical and understandable too.

The fade is choppy, with clear volume steps. I do not recommend fading sounds quickly, slowly fading is a much more pleasant sound experience, but again, just having this is an immense step forward.

Couldn't crash it either :D

:+1:

paramat commented:
bd921a7916f0fafc493b1c4d0eeb5e2bb1d6a7c2
